runtime/js-compute-runtime/builtins/crypto-algorithm.cpp

```diff
@@ -1105,6 +1221,260 @@ JSObject *CryptoAlgorithmHMAC_Import::toObject(JSContext *cx) {
+            case NamedCurve::P521: {
+              if (d.length() != 66) {
+                std::string message = "The JWK's \"d\" member defines an octet string of length ";
+                message += std::to_string(d.length());
+                message += " bytes but should be 66";
+                DOMException::raise(cx, message, "SyntaxError");
+                return nullptr;
+              }
+              break;
+            }
+          }
+          // 2.9.4.2. Let key be a new CryptoKey object that represents the Elliptic Curve private key identified by interpreting jwk according to Section 6.2.2 of JSON Web Algorithms.
+          // 2.9.4.3. Set the [[type]] internal slot of Key to "private".
+          key = createECPrivateKeyFromJWK(cx, jwk);
+          if (!key) {
+            return nullptr;
+          }
+        // Otherwise:
+        } else {
+          // 2.9.4.1. If jwk does not meet the requirements of Section 6.2.1 of JSON Web Algorithms, then throw a DataError.
+          // 2.9.4.2. Let key be a new CryptoKey object that represents the Elliptic Curve public key identified by interpreting jwk according to Section 6.2.1 of JSON Web Algorithms.
+          // 2.9.4.3. Set the [[type]] internal slot of Key to "public".
+          key = createECPublicKeyFromJWK(cx, jwk);
+          if (!key) {
+            return nullptr;
+          }
+        }
+      //  Otherwise:
+        // 2.9.1. Perform any key import steps defined by other applicable specifications, passing format, jwk and obtaining key.
+        // 2.9.2. If an error occured or there are no applicable specifications, throw a DataError.
+        // NOTE: We do not implement the above 2 steps as we only support "P-256", "P-384", and "P-521".
+
+      // 2.10. If the key value is not a valid point on the Elliptic Curve identified by the namedCurve member of normalizedAlgorithm throw a DataError.
+      // 2.11. Let algorithm be a new instance of an EcKeyAlgorithm object.
+      // 2.12. Set the name attribute of algorithm to "ECDSA".
+      // 2.13. Set the namedCurve attribute of algorithm to namedCurve.
+      // 2.14. Set the [[algorithm]] internal slot of key to algorithm.
+      return CryptoKey::createECDSA(cx, this, std::move(key), extractable, usages);
+      break;
+    }
+    case CryptoKeyFormat::Pkcs8:
+    case CryptoKeyFormat::Raw:
+    case CryptoKeyFormat::Spki: {
+      MOZ_ASSERT(false);
+      // TODO finish this
+    }
+  }
+  return nullptr;
+}
+
+JSObject *CryptoAlgorithmECDSA_Import::importKey(JSContext *cx, CryptoKeyFormat format, JS::HandleValue key_data, bool extractable,
+                      CryptoKeyUsages usages) {
+  MOZ_ASSERT(cx);
+
+  KeyData data;
+  switch (format) {
+    case CryptoKeyFormat::Jwk: {
+      // This handles step 2.3: If the "kty" field of jwk is not "EC", then throw a DataError.
+      auto jwk = JsonWebKey::parse(cx, key_data, "EC");
+      if (!jwk) {
+        return nullptr;
+      }
+      data = jwk.release();
```

Comment:
No need to change it in this pr, but what do you think about making `KeyData` into variant that holds a `std::unique_ptr<JsonWebKey>` instead of a raw pointer? Otherwise we have to remember to free the key at some point in the future.